// eslint-plugin-jsx-a11y/lib/rules/aria-activedescendant-has-tabindex.js


const _ariaQuery = require('aria-query');

const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

const _getTabIndex = require('../util/getTabIndex');

const _getTabIndex2 = _interopRequireDefault(_getTabIndex);

const _isInteractiveElement = require('../util/isInteractiveElement');

const _isInteractiveElement2 = _interopRequireDefault(_isInteractiveElement);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } return Array.from(arr); } /**
                                                                                                                                                                                                     * @fileoverview Enforce elements with aria-activedescendant are tabbable.
                                                                                                                                                                                                     * @author Jesse Beach <@jessebeach>
                                                                                                                                                                                                     */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = 'An element that manages focus with `aria-activedescendant` must be tabbable';

const schema = (0, _schemas.generateObjSchema)();

const domElements = [].concat(_toConsumableArray(_ariaQuery.dom.keys()));

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const attributes = node.attributes;


        if ((0, _jsxAstUtils.getProp)(attributes, 'aria-activedescendant') === undefined) {
          return;
        }

        const type = (0, _jsxAstUtils.elementType)(node);
        // Do not test higher level JSX components, as we do not know what
        // low-level DOM element this maps to.
        if (domElements.indexOf(type) === -1) {
          return;
        }
        const tabIndex = (0, _getTabIndex2.default)((0, _jsxAstUtils.getProp)(attributes, 'tabIndex'));

        // If this is an interactive element, tabIndex must be either left
        // unspecified allowing the inherent tabIndex to obtain or it must be
        // zero (allowing for positive, even though that is not ideal). It cannot
        // be given a negative value.
        if ((0, _isInteractiveElement2.default)(type, attributes) && (tabIndex === undefined || tabIndex >= 0)) {
          return;
        }

        if (tabIndex >= 0) {
          return;
        }

        context.report({
          node,
          message: errorMessage,
        });
      },
    };
  },
};
