// eslint-plugin-jsx-a11y/lib/rules/html-has-lang.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce html element has lang prop.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = '<html> elements must have the lang prop.';

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const type = (0, _jsxAstUtils.elementType)(node);

        if (type && type !== 'html') {
          return;
        }

        const lang = (0, _jsxAstUtils.getPropValue)((0, _jsxAstUtils.getProp)(node.attributes, 'lang'));

        if (lang) {
          return;
        }

        context.report({
          node,
          message: errorMessage,
        });
      },
    };
  },
};
