// eslint-plugin-jsx-a11y/lib/rules/iframe-has-title.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce iframe elements have a title attribute.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = '<iframe> elements must have a unique title property.';

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const type = (0, _jsxAstUtils.elementType)(node);

        if (type && type !== 'iframe') {
          return;
        }

        const title = (0, _jsxAstUtils.getPropValue)((0, _jsxAstUtils.getProp)(node.attributes, 'title'));

        if (title && typeof title === 'string') {
          return;
        }

        context.report({
          node,
          message: errorMessage,
        });
      },
    };
  },
};
