// eslint-plugin-jsx-a11y/lib/rules/interactive-supports-focus.js


const _ariaQuery = require('aria-query');

const _jsxAstUtils = require('jsx-ast-utils');

const _arrayIncludes = require('array-includes');

const _arrayIncludes2 = _interopRequireDefault(_arrayIncludes);

const _schemas = require('../util/schemas');

const _isHiddenFromScreenReader = require('../util/isHiddenFromScreenReader');

const _isHiddenFromScreenReader2 = _interopRequireDefault(_isHiddenFromScreenReader);

const _isInteractiveElement = require('../util/isInteractiveElement');

const _isInteractiveElement2 = _interopRequireDefault(_isInteractiveElement);

const _isInteractiveRole = require('../util/isInteractiveRole');

const _isInteractiveRole2 = _interopRequireDefault(_isInteractiveRole);

const _isNonInteractiveElement = require('../util/isNonInteractiveElement');

const _isNonInteractiveElement2 = _interopRequireDefault(_isNonInteractiveElement);

const _isNonInteractiveRole = require('../util/isNonInteractiveRole');

const _isNonInteractiveRole2 = _interopRequireDefault(_isNonInteractiveRole);

const _isPresentationRole = require('../util/isPresentationRole');

const _isPresentationRole2 = _interopRequireDefault(_isPresentationRole);

const _getTabIndex = require('../util/getTabIndex');

const _getTabIndex2 = _interopRequireDefault(_getTabIndex);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } return Array.from(arr); } /**
                                                                                                                                                                                                     * @fileoverview Enforce that elements with onClick handlers must be tabbable.
                                                                                                                                                                                                     * @author Ethan Cohen
                                                                                                                                                                                                     *
                                                                                                                                                                                                     */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const schema = (0, _schemas.generateObjSchema)({
  tabbable: (0, _schemas.enumArraySchema)([].concat(_toConsumableArray(_ariaQuery.roles.keys())).filter(name => !_ariaQuery.roles.get(name).abstract).filter(name => _ariaQuery.roles.get(name).superClass.some(klasses => (0, _arrayIncludes2.default)(klasses, 'widget')))),
});
const domElements = [].concat(_toConsumableArray(_ariaQuery.dom.keys()));

const interactiveProps = [].concat(_toConsumableArray(_jsxAstUtils.eventHandlersByType.mouse), _toConsumableArray(_jsxAstUtils.eventHandlersByType.keyboard));

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const tabbable = context.options && context.options[0] && context.options[0].tabbable || [];
        const attributes = node.attributes;

        const type = (0, _jsxAstUtils.elementType)(node);
        const hasInteractiveProps = (0, _jsxAstUtils.hasAnyProp)(attributes, interactiveProps);
        const hasTabindex = (0, _getTabIndex2.default)((0, _jsxAstUtils.getProp)(attributes, 'tabIndex')) !== undefined;

        if (!(0, _arrayIncludes2.default)(domElements, type)) {
          // Do not test higher level JSX components, as we do not know what
          // low-level DOM element this maps to.
          return;
        } else if (!hasInteractiveProps || (0, _isHiddenFromScreenReader2.default)(type, attributes) || (0, _isPresentationRole2.default)(type, attributes)) {
          // Presentation is an intentional signal from the author that this
          // element is not meant to be perceivable. For example, a click screen
          // to close a dialog .
          return;
        }

        if (hasInteractiveProps && (0, _isInteractiveRole2.default)(type, attributes) && !(0, _isInteractiveElement2.default)(type, attributes) && !(0, _isNonInteractiveElement2.default)(type, attributes) && !(0, _isNonInteractiveRole2.default)(type, attributes) && !hasTabindex) {
          const role = (0, _jsxAstUtils.getLiteralPropValue)((0, _jsxAstUtils.getProp)(attributes, 'role'));
          if ((0, _arrayIncludes2.default)(tabbable, role)) {
            // Always tabbable, tabIndex = 0
            context.report({
              node,
              message: `Elements with the '${role}' interactive role must be tabbable.`,
            });
          } else {
            // Focusable, tabIndex = -1 or 0
            context.report({
              node,
              message: `Elements with the '${role}' interactive role must be focusable.`,
            });
          }
        }
      },
    };
  },
};
