// eslint-plugin-jsx-a11y/lib/rules/label-has-for.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

const _hasAccessibleChild = require('../util/hasAccessibleChild');

const _hasAccessibleChild2 = _interopRequireDefault(_hasAccessibleChild);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const errorMessage = 'Form label must have associated control'; /**
                                                               * @fileoverview Enforce label tags have htmlFor attribute.
                                                               * @author Ethan Cohen
                                                               */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const enumValues = ['nesting', 'id'];
const schema = {
  type: 'object',
  properties: {
    components: _schemas.arraySchema,
    required: {
      oneOf: [{ type: 'string', enum: enumValues }, (0, _schemas.generateObjSchema)({ some: (0, _schemas.enumArraySchema)(enumValues) }, ['some']), (0, _schemas.generateObjSchema)({ every: (0, _schemas.enumArraySchema)(enumValues) }, ['every'])],
    },
    allowChildren: { type: 'boolean' },
  },
};

const validateNesting = function validateNesting(node) {
  return node.parent.children.some(child => child.type === 'JSXElement');
};

const validateId = function validateId(node) {
  const htmlForAttr = (0, _jsxAstUtils.getProp)(node.attributes, 'htmlFor');
  const htmlForValue = (0, _jsxAstUtils.getPropValue)(htmlForAttr);

  return htmlForAttr !== false && !!htmlForValue;
};

const validate = function validate(node, required, allowChildren) {
  if (allowChildren === true) {
    return (0, _hasAccessibleChild2.default)(node.parent);
  }
  if (required === 'nesting') {
    return validateNesting(node);
  }
  return validateId(node);
};

const isValid = function isValid(node, required, allowChildren) {
  if (Array.isArray(required.some)) {
    return required.some.some(rule => validate(node, rule, allowChildren));
  } else if (Array.isArray(required.every)) {
    return required.every.every(rule => validate(node, rule, allowChildren));
  }

  return validate(node, required, allowChildren);
};

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const options = context.options[0] || {};
        const componentOptions = options.components || [];
        const typesToValidate = ['label'].concat(componentOptions);
        const nodeType = (0, _jsxAstUtils.elementType)(node);

        // Only check 'label' elements and custom types.
        if (typesToValidate.indexOf(nodeType) === -1) {
          return;
        }

        const required = options.required || { every: ['nesting', 'id'] };
        const allowChildren = options.allowChildren || false;

        if (!isValid(node, required, allowChildren)) {
          context.report({
            node,
            message: errorMessage,
          });
        }
      },
    };
  },
};
