// eslint-plugin-jsx-a11y/lib/rules/lang.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

const _ISO = require('../util/attributes/ISO.json');

const _ISO2 = _interopRequireDefault(_ISO);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const errorMessage = 'lang attribute must have a valid value.'; /**
                                                               * @fileoverview Enforce lang attribute has a valid value.
                                                               * @author Ethan Cohen
                                                               */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXAttribute: function JSXAttribute(node) {
        const name = (0, _jsxAstUtils.propName)(node);
        if (name && name.toUpperCase() !== 'LANG') {
          return;
        }

        const parent = node.parent;

        const type = (0, _jsxAstUtils.elementType)(parent);
        if (type && type !== 'html') {
          return;
        }

        const value = (0, _jsxAstUtils.getLiteralPropValue)(node);

        // Don't check identifiers
        if (value === null) {
          return;
        } else if (value === undefined) {
          context.report({
            node,
            message: errorMessage,
          });

          return;
        }

        const hyphen = value.indexOf('-');
        const lang = hyphen > -1 ? value.substring(0, hyphen) : value;
        const country = hyphen > -1 ? value.substring(3) : undefined;

        if (_ISO2.default.languages.indexOf(lang) > -1 && (country === undefined || _ISO2.default.countries.indexOf(country) > -1)) {
          return;
        }

        context.report({
          node,
          message: errorMessage,
        });
      },
    };
  },
};
