// eslint-plugin-jsx-a11y/lib/rules/mouse-events-have-key-events.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce onmouseover/onmouseout are
 *  accompanied by onfocus/onblur.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const mouseOverErrorMessage = 'onMouseOver must be accompanied by onFocus for accessibility.';
const mouseOutErrorMessage = 'onMouseOut must be accompanied by onBlur for accessibility.';

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const attributes = node.attributes;

        // Check onmouseover / onfocus pairing.

        const onMouseOver = (0, _jsxAstUtils.getProp)(attributes, 'onMouseOver');
        const onMouseOverValue = (0, _jsxAstUtils.getPropValue)(onMouseOver);

        if (onMouseOver && onMouseOverValue != null) {
          const hasOnFocus = (0, _jsxAstUtils.getProp)(attributes, 'onFocus');
          const onFocusValue = (0, _jsxAstUtils.getPropValue)(hasOnFocus);

          if (hasOnFocus === false || onFocusValue === null || onFocusValue === undefined) {
            context.report({
              node,
              message: mouseOverErrorMessage,
            });
          }
        }

        // Checkout onmouseout / onblur pairing
        const onMouseOut = (0, _jsxAstUtils.getProp)(attributes, 'onMouseOut');
        const onMouseOutValue = (0, _jsxAstUtils.getPropValue)(onMouseOut);
        if (onMouseOut && onMouseOutValue != null) {
          const hasOnBlur = (0, _jsxAstUtils.getProp)(attributes, 'onBlur');
          const onBlurValue = (0, _jsxAstUtils.getPropValue)(hasOnBlur);

          if (hasOnBlur === false || onBlurValue === null || onBlurValue === undefined) {
            context.report({
              node,
              message: mouseOutErrorMessage,
            });
          }
        }
      },
    };
  },
};
