// eslint-plugin-jsx-a11y/lib/rules/no-autofocus.js


const _jsxAstUtils = require('jsx-ast-utils');

const _ariaQuery = require('aria-query');

const _schemas = require('../util/schemas');

const errorMessage = 'The autoFocus prop should not be used, as it can reduce usability and accessibility for users.'; /**
                                                                                                                      * @fileoverview Enforce autoFocus prop is not used.
                                                                                                                      * @author Ethan Cohen <@evcohen>
                                                                                                                      */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const schema = (0, _schemas.generateObjSchema)({
  ignoreNonDOM: {
    type: 'boolean',
    default: false,
  },
});

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXAttribute: function JSXAttribute(attribute) {
        // Determine if ignoreNonDOM is set to true
        // If true, then do not run rule.
        const options = context.options[0] || {};
        const ignoreNonDOM = !!options.ignoreNonDOM;

        if (ignoreNonDOM) {
          const type = (0, _jsxAstUtils.elementType)(attribute.parent);
          if (!_ariaQuery.dom.get(type)) {
            return;
          }
        }

        // Don't normalize, since React only recognizes autoFocus on low-level DOM elements.
        if ((0, _jsxAstUtils.propName)(attribute) === 'autoFocus') {
          context.report({
            node: attribute,
            message: errorMessage,
          });
        }
      },
    };
  },
};
