// eslint-plugin-jsx-a11y/lib/rules/no-access-key.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce no accesskey attribute on element.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = 'No access key attribute allowed. Inconsistencies ' + 'between keyboard shortcuts and keyboard comments used by screenreader ' + 'and keyboard only users create a11y complications.';

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const accessKey = (0, _jsxAstUtils.getProp)(node.attributes, 'accesskey');
        const accessKeyValue = (0, _jsxAstUtils.getPropValue)(accessKey);

        if (accessKey && accessKeyValue) {
          context.report({
            node,
            message: errorMessage,
          });
        }
      },
    };
  },
};
