// eslint-plugin-jsx-a11y/lib/rules/no-distracting-elements.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce distracting elements are not used.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = function errorMessage(element) {
  return `Do not use <${element}> elements as they can create visual accessibility issues and are deprecated.`;
};

const DEFAULT_ELEMENTS = ['marquee', 'blink'];

const schema = (0, _schemas.generateObjSchema)({
  elements: (0, _schemas.enumArraySchema)(DEFAULT_ELEMENTS),
});

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const options = context.options[0] || {};
        const elementOptions = options.elements || DEFAULT_ELEMENTS;
        const type = (0, _jsxAstUtils.elementType)(node);
        const distractingElement = elementOptions.find(element => type === element);

        if (distractingElement) {
          context.report({
            node,
            message: errorMessage(distractingElement),
          });
        }
      },
    };
  },
};
