// eslint-plugin-jsx-a11y/lib/rules/no-noninteractive-tabindex.js


const _extends = Object.assign || function (target) { for (let i = 1; i < arguments.length; i++) { const source = arguments[i]; for (const key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * @fileoverview Disallow tabindex on static and noninteractive elements
                                                                                                                                                                                                                                                                   * @author jessebeach
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const _ariaQuery = require('aria-query');

const _jsxAstUtils = require('jsx-ast-utils');

const _arrayIncludes = require('array-includes');

const _arrayIncludes2 = _interopRequireDefault(_arrayIncludes);

const _isInteractiveElement = require('../util/isInteractiveElement');

const _isInteractiveElement2 = _interopRequireDefault(_isInteractiveElement);

const _isInteractiveRole = require('../util/isInteractiveRole');

const _isInteractiveRole2 = _interopRequireDefault(_isInteractiveRole);

const _schemas = require('../util/schemas');

const _getTabIndex = require('../util/getTabIndex');

const _getTabIndex2 = _interopRequireDefault(_getTabIndex);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const errorMessage = '`tabIndex` should only be declared on interactive elements.';

const schema = (0, _schemas.generateObjSchema)({
  roles: _extends({}, _schemas.arraySchema, {
    description: 'An array of ARIA roles',
  }),
  tags: _extends({}, _schemas.arraySchema, {
    description: 'An array of HTML tag names',
  }),
});

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    const options = context.options;

    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const type = (0, _jsxAstUtils.elementType)(node);
        const attributes = node.attributes;

        const tabIndexProp = (0, _jsxAstUtils.getProp)(attributes, 'tabIndex');
        const tabIndex = (0, _getTabIndex2.default)(tabIndexProp);
        // Early return;
        if (typeof tabIndex === 'undefined') {
          return;
        }
        const role = (0, _jsxAstUtils.getLiteralPropValue)((0, _jsxAstUtils.getProp)(node.attributes, 'role'));

        if (!_ariaQuery.dom.has(type)) {
          // Do not test higher level JSX components, as we do not know what
          // low-level DOM element this maps to.
          return;
        }
        // Allow for configuration overrides.

        let _ref = options[0] || {},
          tags = _ref.tags,
          roles = _ref.roles;

        if (tags && (0, _arrayIncludes2.default)(tags, type) || roles && (0, _arrayIncludes2.default)(roles, role)) {
          return;
        }
        if ((0, _isInteractiveElement2.default)(type, attributes) || (0, _isInteractiveRole2.default)(type, attributes)) {
          return;
        }
        if (tabIndex >= 0) {
          context.report({
            node: tabIndexProp,
            message: errorMessage,
          });
        }
      },
    };
  },
};
