// eslint-plugin-jsx-a11y/lib/rules/no-onchange.js


const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

/**
 * @fileoverview Enforce usage of onBlur over onChange for accessibility.
 * @author Ethan Cohen
 */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = 'onBlur must be used instead of onchange, ' + 'unless absolutely necessary and it causes no negative consequences ' + 'for keyboard only or screen reader users.';

const applicableTypes = ['select', 'option'];

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        const nodeType = (0, _jsxAstUtils.elementType)(node);

        if (applicableTypes.indexOf(nodeType) === -1) {
          return;
        }

        const onChange = (0, _jsxAstUtils.getProp)(node.attributes, 'onChange');
        const hasOnBlur = (0, _jsxAstUtils.getProp)(node.attributes, 'onBlur') !== undefined;

        if (onChange && !hasOnBlur) {
          context.report({
            node,
            message: errorMessage,
          });
        }
      },
    };
  },
};
