// eslint-plugin-jsx-a11y/lib/rules/role-supports-aria-props.js


const _ariaQuery = require('aria-query');

const _jsxAstUtils = require('jsx-ast-utils');

const _schemas = require('../util/schemas');

const _getImplicitRole = require('../util/getImplicitRole');

const _getImplicitRole2 = _interopRequireDefault(_getImplicitRole);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } return Array.from(arr); } /**
                                                                                                                                                                                                     * @fileoverview Enforce that elements with explicit or implicit roles defined contain only
                                                                                                                                                                                                     * `aria-*` properties supported by that `role`.
                                                                                                                                                                                                     * @author Ethan Cohen
                                                                                                                                                                                                     */

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------

const errorMessage = function errorMessage(attr, role, tag, isImplicit) {
  if (isImplicit) {
    return `The attribute ${attr} is not supported by the role ${role}. This role is implicit on the element ${tag}.`;
  }

  return `The attribute ${attr} is not supported by the role ${role}.`;
};

const schema = (0, _schemas.generateObjSchema)();

module.exports = {
  meta: {
    docs: {},
    schema: [schema],
  },

  create: function create(context) {
    return {
      JSXOpeningElement: function JSXOpeningElement(node) {
        // If role is not explicitly defined, then try and get its implicit role.
        const type = (0, _jsxAstUtils.elementType)(node);
        const role = (0, _jsxAstUtils.getProp)(node.attributes, 'role');
        const roleValue = role ? (0, _jsxAstUtils.getLiteralPropValue)(role) : (0, _getImplicitRole2.default)(type, node.attributes);
        const isImplicit = roleValue && role === undefined;

        // If there is no explicit or implicit role, then assume that the element
        // can handle the global set of aria-* properties.
        // This actually isn't true - should fix in future release.
        if (typeof roleValue !== 'string' || _ariaQuery.roles.get(roleValue) === undefined) {
          return;
        }

        // Make sure it has no aria-* properties defined outside of its property set.

        let _roles$get = _ariaQuery.roles.get(roleValue),
          propKeyValues = _roles$get.props;

        const propertySet = Object.keys(propKeyValues);
        const invalidAriaPropsForRole = [].concat(_toConsumableArray(_ariaQuery.aria.keys())).filter(attribute => propertySet.indexOf(attribute) === -1);

        node.attributes.forEach((prop) => {
          if (prop.type === 'JSXSpreadAttribute') {
            return;
          }

          const name = (0, _jsxAstUtils.propName)(prop);
          if (invalidAriaPropsForRole.indexOf(name) > -1) {
            context.report({
              node,
              message: errorMessage(name, roleValue, type, isImplicit),
            });
          }
        });
      },
    };
  },
};
