// eslint-plugin-jsx-a11y/lib/util/implicitRoles/index.js


Object.defineProperty(exports, '__esModule', {
  value: true,
});

const _a = require('./a');

const _a2 = _interopRequireDefault(_a);

const _area = require('./area');

const _area2 = _interopRequireDefault(_area);

const _article = require('./article');

const _article2 = _interopRequireDefault(_article);

const _aside = require('./aside');

const _aside2 = _interopRequireDefault(_aside);

const _body = require('./body');

const _body2 = _interopRequireDefault(_body);

const _button = require('./button');

const _button2 = _interopRequireDefault(_button);

const _datalist = require('./datalist');

const _datalist2 = _interopRequireDefault(_datalist);

const _details = require('./details');

const _details2 = _interopRequireDefault(_details);

const _dialog = require('./dialog');

const _dialog2 = _interopRequireDefault(_dialog);

const _dl = require('./dl');

const _dl2 = _interopRequireDefault(_dl);

const _form = require('./form');

const _form2 = _interopRequireDefault(_form);

const _h = require('./h1');

const _h2 = _interopRequireDefault(_h);

const _h3 = require('./h2');

const _h4 = _interopRequireDefault(_h3);

const _h5 = require('./h3');

const _h6 = _interopRequireDefault(_h5);

const _h7 = require('./h4');

const _h8 = _interopRequireDefault(_h7);

const _h9 = require('./h5');

const _h10 = _interopRequireDefault(_h9);

const _h11 = require('./h6');

const _h12 = _interopRequireDefault(_h11);

const _hr = require('./hr');

const _hr2 = _interopRequireDefault(_hr);

const _img = require('./img');

const _img2 = _interopRequireDefault(_img);

const _input = require('./input');

const _input2 = _interopRequireDefault(_input);

const _li = require('./li');

const _li2 = _interopRequireDefault(_li);

const _link = require('./link');

const _link2 = _interopRequireDefault(_link);

const _menu = require('./menu');

const _menu2 = _interopRequireDefault(_menu);

const _menuitem = require('./menuitem');

const _menuitem2 = _interopRequireDefault(_menuitem);

const _meter = require('./meter');

const _meter2 = _interopRequireDefault(_meter);

const _nav = require('./nav');

const _nav2 = _interopRequireDefault(_nav);

const _ol = require('./ol');

const _ol2 = _interopRequireDefault(_ol);

const _option = require('./option');

const _option2 = _interopRequireDefault(_option);

const _output = require('./output');

const _output2 = _interopRequireDefault(_output);

const _progress = require('./progress');

const _progress2 = _interopRequireDefault(_progress);

const _section = require('./section');

const _section2 = _interopRequireDefault(_section);

const _select = require('./select');

const _select2 = _interopRequireDefault(_select);

const _tbody = require('./tbody');

const _tbody2 = _interopRequireDefault(_tbody);

const _textarea = require('./textarea');

const _textarea2 = _interopRequireDefault(_textarea);

const _tfoot = require('./tfoot');

const _tfoot2 = _interopRequireDefault(_tfoot);

const _thead = require('./thead');

const _thead2 = _interopRequireDefault(_thead);

const _ul = require('./ul');

const _ul2 = _interopRequireDefault(_ul);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  a: _a2.default,
  area: _area2.default,
  article: _article2.default,
  aside: _aside2.default,
  body: _body2.default,
  button: _button2.default,
  datalist: _datalist2.default,
  details: _details2.default,
  dialog: _dialog2.default,
  dl: _dl2.default,
  form: _form2.default,
  h1: _h2.default,
  h2: _h4.default,
  h3: _h6.default,
  h4: _h8.default,
  h5: _h10.default,
  h6: _h12.default,
  hr: _hr2.default,
  img: _img2.default,
  input: _input2.default,
  li: _li2.default,
  link: _link2.default,
  menu: _menu2.default,
  menuitem: _menuitem2.default,
  meter: _meter2.default,
  nav: _nav2.default,
  ol: _ol2.default,
  option: _option2.default,
  output: _output2.default,
  progress: _progress2.default,
  section: _section2.default,
  select: _select2.default,
  tbody: _tbody2.default,
  textarea: _textarea2.default,
  tfoot: _tfoot2.default,
  thead: _thead2.default,
  ul: _ul2.default,
};
