// eslint-plugin-jsx-a11y/lib/util/schemas.js


Object.defineProperty(exports, '__esModule', {
  value: true,
});
/**
 * JSON schema to accept an array of unique strings
 */
const arraySchema = exports.arraySchema = {
  type: 'array',
  items: {
    type: 'string',
  },
  uniqueItems: true,
  additionalItems: false,
};

/**
 * JSON schema to accept an array of unique strings from an enumerated list.
 */
const enumArraySchema = exports.enumArraySchema = function enumArraySchema() {
  const enumeratedList = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  const minItems = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return Object.assign({}, arraySchema, {
    items: {
      type: 'string',
      enum: enumeratedList,
    },
    minItems,
  });
};

/**
 * Factory function to generate an object schema
 * with specified properties object
 */
const generateObjSchema = exports.generateObjSchema = function generateObjSchema() {
  const properties = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const required = arguments[1];
  return {
    type: 'object',
    properties,
    required,
  };
};
